lib/jwt/decode.rb

```diff
@@ -152,15 +153,35 @@ def header
     end
 
     def payload
-      @payload ||= parse_and_decode @segments[1]
+      @payload ||= parse_and_decode(encoded_payload, decode: decode_payload?)
+    end
+
+    def encoded_payload
+      payload = encoded_detached_payload if !@detached_payload.nil? && @segments[1].empty?
```

Comment:
I think it would make sense to raise an error if a payload is given and the token also contains a payload, instead of just ignoring the given payload.